Display Order Items and Shipping Labels Packages as separate sections in the Order Detail Screen

pmusolino opened the pull request:
Fixes #4436  
Fixes #4412

## Description
In this PR, I tried to fix two different issues that we encountered in the Order Detail screen.
The first issue was about displaying the Products section and the Shipping Label Packages sections at the same time, because previously, if there was an existing package, replaced the products. This change will allow us to solve this issue in the future https://github.com/woocommerce/woocommerce-ios/issues/4126 and this one #4412, which I solved following the code implemented previously by @koke here #4416.
This implementation, changed a little bit the Shipping Label Package section because now the items are not showed inside the section, but by tapping on them you will see a new screen showing the package items.
I also removed the cell "Details" under Products, because it was superfluous.

## Testing
#### Case 1
1. Open an order detail with partially refunded items.
2. Make sure that everything appears like before.

#### Case 2
1. Open an order, that is eligible for Shipping Label Creation.
2. Make sure to see the shipping label creation button under the Order Items.

#### Case 3
1. Open an order, with existing Shipping Label Packages.
2. Make sure you are able to see the order items in the order detail screen, without any button below (for example, you shouldn't see the shipping label creation button).
3. Make sure that you are able to see the shipping label packages under the order items section.

#### Case 4
1. Create a variable product.
2. Add “Color” as an attribute with “Blue|Green|Yellow” as the values.
3. Create a single variation of that variable product. Choose “Any” as the attribute option.
4. Create a single order for the product.
5. Add 1 Blue
6. Add 2 Green
7. Add 5 Yellow
8. Check out and submit the order.
9. In the app, view the order. It should show three different items, one for each color purchased.


## Screenshots
| New Shipping Label Package section in Order Detail             |  Shipping Label Package Items detail screen |
:-------------------------:|:-------------------------:
![Simulator Screen Shot - iPhone 12 Pro - 2021-06-17 at 13 33 42](https://user-images.githubusercontent.com/495617/122403011-49727980-cf7e-11eb-916f-cec89c283155.png) | ![Simulator Screen Shot - iPhone 12 Pro - 2021-06-17 at 13 33 49](https://user-images.githubusercontent.com/495617/122403024-4bd4d380-cf7e-11eb-8f92-cfc3f553a921.png)

Variable Products Before | Variable Products Before After
-|-
![Screen Shot 2021-06-15 at 09 44 15](https://user-images.githubusercontent.com/8739/122014512-83495180-cdbf-11eb-9d9a-631d9a417643.png)|![Screen Shot 2021-06-15 at 09 44 59](https://user-images.githubusercontent.com/8739/122014520-85131500-cdbf-11eb-98fe-292980a40158.png)



Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

pmusolino committed:
feat: removed "Details" cell in Order Details, since the data showed inside are the same of the order items. Now we always display the products also if there are some the existing packages from shipping labels.

pmusolino committed:
feat: Include attributes for aggregating order items

pmusolino committed:
feat: in Order Detail now we display the packages as separate sections, and there will be inside every section a button that open the items list of the specified package.

pmusolino committed:
update: RELEASE-NOTES

pmusolino committed:
fix: sort order in RELEASE-NOTES

pmusolino committed:
fix: add-on workaround relies on that item order to properly crossreference add-ons

Ecarrion commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift:
```diff
@@ -571,23 +564,33 @@ private extension OrderDetailsDataSource {
-            self?.onCellAction?(.viewAddOns(addOns: addOns), nil)
-        }
+        let orderItems = shippingLabelOrderItemsAggregator.orderItems(of: shippingLabel)
+        let singular = NSLocalizedString("%d item",
+                                         comment: "For example: `1 item` in Shipping Label package row")
+        let plural = NSLocalizedString("%d items",
+                                       comment: "For example: `5 items` in Shipping Label package row")
```
nit: I think we should use `%1$@` in those formats.

Ecarrion commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift:
```diff
@@ -571,23 +564,33 @@ private extension OrderDetailsDataSource {
+                                         comment: "For example: `1 item` in Shipping Label package row")
+        let plural = NSLocalizedString("%d items",
+                                       comment: "For example: `5 items` in Shipping Label package row")
+        let itemsText = String.pluralize(orderItems.count, singular: singular, plural: plural)
```
TIL

Ecarrion commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift:
```diff
@@ -258,28 +258,21 @@ extension OrderDetailsViewModel {
+            let shippingLabelItems = dataSource.shippingLabelOrderItems(at: indexPath)
+
+            let identifier = AggregatedProductListViewController.classNameWithoutNamespaces
+            guard let productListVC = UIStoryboard.orders.instantiateViewController(identifier: identifier) as? AggregatedProductListViewController else {
+                DDLogError("Error: attempted to instantiate AggregatedProductListViewController. Instantiation failed.")
```
I was under the impression that we are moving away from Storyboards. What was the reason to use it for this screen?

pmusolino replied:
The reason is that the similar view controller that works with `OrderItem` instead of `AggregatedOrderItem` is in the storyboard. I, therefore, preferred to conform and use the same approach by inserting it in the storyboard.

Ecarrion commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift:
```diff
@@ -258,28 +258,21 @@ extension OrderDetailsViewModel {
-            viewController.present(navController, animated: true, completion: nil)
+            productListVC.viewModel = self
+            productListVC.items = shippingLabelItems
+            viewController.navigationController?.show(productListVC, sender: nil)
```
nit: I think it can be `viewController.how(productListVC, sender: nil)`

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Product List Section/Product Details/AggregatedProductListViewController.swift:
```diff
@@ -0,0 +1,148 @@
+    private var products: [Product]? = []
+    var items: [AggregateOrderItem] = []
+
+    private let imageService: ImageService = ServiceLocator.imageService
```
I really think we need to turn this into a normal VC with nib, to be able to inject these variables.

pmusolino replied:
I'm going to convert it into xib 👍 I did it just to conform to the existing view controllers that are in the `Orders` storyboard.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Product List Section/Product Details/AggregatedProductListViewController.swift:
```diff
@@ -0,0 +1,148 @@
+        return items[indexPath.row]
+    }
+
+    func lookUpProduct(by productID: Int64) -> Product? {
+        return products?.filter({ $0.productID == productID }).first
+    }
```
I think this logic should be inside the view model. That would allow us to add some tests to it.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Product List Section/Product Details/AggregatedProductListViewController.swift:
```diff
@@ -0,0 +1,148 @@
+    /// Returns the item attributes that can be identified as add-ons attributes.
+    /// If the "view add-ons" feature is disabled an empty array will be returned.
+    ///
+    func itemAddOnsAttributes(item: AggregateOrderItem) -> [OrderItemAttribute] {
+        guard let product = lookUpProduct(by: item.productID), viewModel.dataSource.showAddOns else {
+            return []
+        }
+
+        let globalAddOns = viewModel.dataSource.addOnGroups
+        return AddOnCrossreferenceUseCase(orderItemAttributes: item.attributes, product: product, addOnGroups: globalAddOns).addOnsAttributes()
+    }
```
I think we should not show the "View Add-ons" button on the shipped items screen because we can't really know if a shipped item contained an add-on.

Could we remove all the code associated with it?

<img width="404" alt="Screen Shot 2021-06-17 at 8 43 57 AM" src="https://user-images.githubusercontent.com/562080/122411644-87ee4100-cf4a-11eb-85d4-da2b90596d7b.png">

pmusolino replied:
Removed 👍

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Product List Section/Product Details/AggregatedProductListViewController.swift:
```diff
@@ -0,0 +1,148 @@
+    /// Setup: Main View
+    ///
+    func configureMainView() {
+        title = String.pluralize(items.count, singular: Localization.titleSingular, plural: Localization.titlePlural)
```
Is it ok to group the items count in the here and in package cell, when we actually shipped more items? 😅 

4 items | 22 items
---- | ----
<img width="421" alt="Screen Shot 2021-06-17 at 8 46 09 AM" src="https://user-images.githubusercontent.com/562080/122411873-bbc96680-cf4a-11eb-929c-e3b9639e80de.png"> | <img width="401" alt="Screen Shot 2021-06-17 at 8 46 19 AM" src="https://user-images.githubusercontent.com/562080/122411884-bcfa9380-cf4a-11eb-9f95-42f75aef760d.png">

pmusolino replied:
I'm not sure about the approach that we want to use here. cc @Garance91540 @adamzelinski 
By the way, I added this header on top of the detail screen, because it's probably not clear what the number on the right indicates.

<img src="https://user-images.githubusercontent.com/495617/122424364-f35a0200-cf8e-11eb-82d4-0a6bd3a6296b.png" width=300 />

Garance91540 replied:
Hi @pmusolino let's display the total number of items in the Header (so if the package has 3 of the same product, it should read "3 items"). And good idea for adding "qty", it's clearer that way!

pmusolino replied:
Thank you @Garance91540, I updated the logic 👍

Ecarrion reviewed (changes requested):
Hi Paolo! Thanks for tackling this!

Left some comments, but I'm mostly worried about using storyboards & the "view add-on" button!

pmusolino committed:
fix: moved AggregatedProductListViewController in .xib file, and improved code style + small fixes

pmusolino commented:
Thank you @Ecarrion for the review! Ready for a second pass. I fixed all the things suggested, and moved the `AggregatedProductListViewController` under a xib file.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Product List Section/Product Details/AggregatedProductListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+    @IBOutlet private var tableView: UITableView!
+
+    private var viewModel: OrderDetailsViewModel
+    private var products: [Product]? = []
```
I don't think we need this anymore, do we?

pmusolino replied:
Good catch, removed 👍

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Product List Section/Product Details/AggregatedProductListViewController.swift:
```diff
@@ -0,0 +1,155 @@
+    private var products: [Product]? = []
+    private var items: [AggregateOrderItem]
+
+    private let imageService: ImageService = ServiceLocator.imageService
```
nit: A small improvement would be to inject this with a default parameter, so we make dependencies clear.

pmusolino replied:
Done ✅

Ecarrion reviewed (approved):
Looks good 🚀 
Just some small nits!

Garance91540 reviewed (approved):
👏

pmusolino committed:
fix: logic of shipping labels items count + minor code style fixes
